Implementation and unit tests (milestones 1-4)

davidhazama opened the pull request:
# What

This initial implementation aims to accomplish the first 4 milestones of the integration:
### 1. Initialization
* `sfcode` is checked in the settings and set to `us` per Nielsen documentation
* `appid` is assumed in the settings per Segment documentation
* `appname` is pulled from the application
* `appname` is pulled from the application
* The factory supports a maximum of 4 initializations for a given `appId`
### 2. Supporting started / resumed content
* The following events are supported per the requirements
** Video Content Started
** Video Playback Resumed
** Video Playback Seek Completed
** Video Playback Buffer Completed
* Unit tests for the above events
### 3. Supporting paused content
* The following events are supported per the requirements
** Video Playback Paused
** Video Playback Interrupted
** Video Playback Buffer Started
** Video Playback Seek Started
** Video Content Completed
** Application Backgrounded
* Unit tests for the above events
### 4. Supporting completed content
* The following event is supported per the requirements
** Video Playback Completed
* Unit test for the above event

# Pending Items / Considerations

1. Moving `sendID3` call out of content start / resume handlers and through to appropriate location (ie leaving the `sendID3Block` function to be left to the developer to retrieve)

davidhazama committed:
Initial commit for integration SDK with readme, license, podspec, and unit tests set up

davidhazama committed:
Update README with recommended approach for installing Nielsen App SDK

Update integration with optOutURL and exposing block for usage by client application to hook into timed metadata events
Update integration with tests for the above items

davidhazama committed:
Supporting legacy opt-out method by exposing method to pass opt out URL detected through to Nielsen SDK

davidhazama committed:
Review: Add few more test cases to capture other flows in code

davidhazama committed:
Review: Cleaning up code, updating README

davidhazama committed:
Review: Removing unnecessary class declaration

davidhazama committed:
Review: Updating factory constructor to follow same convention as integration

davidhazama committed:
Update README

briemcnally commented on Segment-Nielsen-DTVR/Classes/SEGNielsenDTVRIntegration.m:
```diff
@@ -0,0 +1,204 @@
+        NSString *appId = settings[@"appId"];
+        
+        NSString *sfcode;
+        if (settings[@"sfcode"]) {
```
Let's change this to `if (!settings[@"sfcode"])` then we will default to `us`.

davidhazama replied:
Done :+1:

briemcnally commented on Segment-Nielsen-DTVR/Classes/SEGNielsenDTVRIntegration.m:
```diff
@@ -0,0 +1,204 @@
+-(void)setupEventHandlers
+{
+    self.eventHandlers = [[NSMutableArray alloc] init];
+    SEGNielsenEventHandler *startHandler = [[SEGNielsenEventHandler alloc]
```
Similar to our comments on android can we add descriptions of the handlers with: 
- @param [param] [Description]
- @return [Description]
This will help other contributors if the code gets complicated down the line.

davidhazama replied:
Done :+1: let me know if on the constructor is sufficient

briemcnally reviewed (commented):
Overall looks great! Good job! 

Per our meeting sync today we will follow up on if we can call sendId3 as a new event rather than real time. We will also follow up on whether or not we can use the onReady method to expose the App SDK to the customer.

Ensemble Action Items:
- Call sendId3 only when the tags change between events.

davidhazama committed:
PR Review: Using 'sfcode' in settings and defaulting to 'us' if none provided

davidhazama committed:
PR Review: Adding headerdoc information to event handler private class

davidhazama commented:
Thank you! Will be working on ensuring the integration will ensure that `sendID3` is only called when new tags are detected

davidhazama committed:
PR Review: Ensuring sendID3 is only called when we see a new ID3 tag on exposed function block

davidhazama committed:
Review: Adding documentation for sendID3 block

davidhazama committed:
PR Review: Adding documentation to other methods in implementation

davidhazama committed:
PR Review: Adding documentation and minor refactor for integration factory

davidhazama committed:
Review: Ensuring valid key access for dictionary if appId is not provided in settings

davidhazama committed:
Updating sendID3 to use event names from destination settings, as well as retrieve ID3 property key in analytics payload properties (which defaults to 'Id3')

Set up tests for above changes

Removed sendID3 block exposure

davidhazama committed:
Removing unused test

davidhazama committed:
Ensuring empty array in case settings are not properly populated

davidhazama committed:
Updating implementation for factory to limit to 4 and simply return old instance if create is called again, rather than create a new instance

davidhazama committed:
Ensuring sendID3 is only called with new ID3 tags, writing tests to verify this

davidhazama committed:
PR Review:

Changing `Id3` default to `id3` default key for ID3 tag

Changing implementation of handling events to ignore letter casing for event names

davidhazama commented:
@briemcnally 

Changes have been made from discussions last week:
- Ensuring sendID3 is only called when new tags are encountered
- sendID3 is now handled as separate event, where event names triggering this tracking call are configured in the destination settings, and the id3 property key to use in analytics events is configurable in destination settings and defaults to `id3` when not specified. Additionally, the event names are not case sensitive
- When a 5th integration is intended to be created, we instead return the first integration in the list

In the sample application:
- Retrieving opt-out URL from integration instance when SEGAnalayticsIntegrationDidStart is fired, and ensuring when the DTVR integration has been created that we retrieve it from the singleton factory. The factory exposes the list of integrations by app ID, and from there we can retrieve the integration to acquire the opt-out URL that can then be displayed in a web view

davidhazama committed:
Listening for 'Debug' flag in settings and enabling logging in Nielsen SDK if present

davidhazama committed:
Fixing debug key for settings

davidhazama committed:
Fixing boolean extraction in settings dictionary
